### PR TITLE
Implement gRPC for TizenRT

### PR DESCRIPTION
**Pull request by an4967**
Only test with **gcc-arm-none-eabi-6-2017-q1-update** [**[download]**](https://developer.arm.com/-/media/Files/downloads/gnu-rm/6_1-2017q1/gcc-arm-none-eabi-6-2017-q1-update-linux.tar.bz2?revision=6799a2bc-ea25-4e11-8a36-3f4bce1c06f7?product=GNU%20Arm%20Embedded%20Toolchain,64-bit,,Linux,6-2017-q1-update)

- Implement sendmsg/recvmsg API
- Add cxxinitialize for cpp build
- Porting protobuf to TizenRT
- Porting gRPC to TizenRT
- Add examples for protobuf client Test
- Add examples for gRPC client Test
- Add configs for gRPC build (artik053/grpc , qemu/tc_16m_grpc)
- Change SSL library (OpenSSL -> mbedTLS)

**How To Test**

0) Check ARM gcc Version (should be over 6.3)
`$ arm-none-eabi-g++ --version`

1) Use artik053/grpc
`os/tools$ ./configure.sh artik053/grpc`

2) Connect to WiFi on TASH (After flashing tinyara.bin)
`TASH>> wifi startsta`
`TASH>> wifi join [SSID] [PASSPHASE]`
`TASH>> ifconfig wl1 dhcp`

3) Run ntpclient (to sync up system time)
`TASH>> ntpclient start 600 0.asia.pool.ntp.org`
`TASH>> date`

4) Run grpc_client or route_client (before this, the grpc server should be running)
To enable secure mode, add [-s] option to the command.
`TASH>> grpc_client [-s] [server ip]:[port]`
`TASH>> route_client [-s] [server ip]:[port]`

**Example Test Environment with Linux**
1)  [TizenRT][Client] **grpc_client** <-> **greeter_server** [Server][Linux]
2) [TizenRT][Client] **route_client** <-> **route_guide_server** [Server][Linux]

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/route_client/helper.h:26: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:27: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:28: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:29: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:30: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:31: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:32: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:33: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:34: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:35: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:36: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:37: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:38: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:39: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:40: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:41: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:42: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:43: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:44: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:45: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:46: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:47: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:48: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:49: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:50: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:51: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:52: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:53: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:54: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:55: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:56: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:57: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:58: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:59: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:60: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:61: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:62: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:63: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:64: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:65: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:66: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:67: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:68: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:69: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:70: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:71: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:72: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:73: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:74: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:75: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:76: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:77: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:78: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:79: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:80: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:81: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:82: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:83: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:84: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:85: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:86: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:87: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:88: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:89: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:90: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:91: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:92: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:93: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:94: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:95: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:96: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:97: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:98: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:99: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:100: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:101: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:102: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:103: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:104: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:105: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:106: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:107: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:108: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:109: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:110: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:111: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:112: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:113: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:114: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:115: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:116: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:117: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:118: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:119: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:120: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:121: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/route_client/helper.h:731: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:VxO)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxO)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] space required before that '*' (ctx:OxW)
apps/examples/route_client/helper.h:733: ERROR: [SPC_M_OPR] space prohibited after that '*' (ctx:OxW)

File: apps/examples/greeter_client/Make.defs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+###########################################################################
+#
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/greeter_client/Make.defs]: check the year

File: apps/examples/greeter_client/greeter_client.cc [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+ * limitations under the License.
+ *
+ */
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <grpc++/grpc++.h>
+#include <tinyara/init.h>
```
Comment (jeongchanKim) [inline apps/examples/greeter_client/greeter_client.cc]: If you modified, Samsung copyright is needed

File: apps/examples/greeter_client/greeter_client.cc [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+};
+
+int greeter_client_cb(void *args) {
+    // Instantiate the client. It requires a channel, out of which the actual RPCs
+    // are created. This channel models a connection to an endpoint (in this case,
+    // localhost at port 50051). We indicate that the channel isn't authenticated
+    // (use of InsecureChannelCredentials()).
+
+    int argc;
+	char **argv;
```
Comment (jeongchanKim) [inline apps/examples/greeter_client/greeter_client.cc]: remove unnecessary space

File: apps/examples/route_client/Makefile [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+-include $(TOPDIR)/Make.defs
+include $(APPDIR)/Make.defs
+
+CXXEXT ?= .cc
+PROTOEXT ?= .proto
+GENCXXEXT ?= .pb.cc
+SERVICECXXEXT ?= .grpc.pb.cc
+MSG_PREFIX ?= .pb
+SERVICE_PREFIX ?= .grpc.pb
+# Protocol Buffers Example
```
Comment (jeongchanKim) [inline apps/examples/route_client/Makefile]: please check this comment

File: apps/examples/route_client/route_guide_client.cc [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+		}
+	} else {
+		printf("Usage: \tgrpc_client <option> [IP:PORT]\n");
+		printf("\t<option>\t-s : secure mode\n");
+		printf("\t<example>\tgrpc_client -s 192.168.1.2:50051\n");
+		return 0;
+	}
+
+	std::shared_ptr<Channel> myChannel;
+#ifdef CONFIG_FS_ROMFS
```
Comment (jeongchanKim) [inline apps/examples/route_client/route_guide_client.cc]: #include<tinyara/config.h> is needed

File: build/configs/qemu/tc_16m_grpc/Make.defs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+############################################################################
+# build/configs/qemu/tc_16m/Make.defs
```
Comment (jeongchanKim) [inline build/configs/qemu/tc_16m_grpc/Make.defs]: please check the path.

File: external/grpc/extlib/zlib/deflate.h [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+#define L_CODES (LITERALS+1+LENGTH_CODES)
+/* number of Literal or Length codes, including the END_BLOCK code */
+
+#define D_CODES   30
+/* number of distance codes */
+
+#define BL_CODES  19
+/* number of codes used to transfer the bit lengths */
+
+#define HEAP_SIZE (2*L_CODES+1)
```
Comment (jeongchanKim) [inline external/grpc/extlib/zlib/deflate.h]: add space between operator and operand

File: external/grpc/extlib/zlib/deflate.h [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+    uInt good_match;
+    /* Use a faster search when the previous match is longer than this */
+
+    int nice_match; /* Stop searching when current match exceeds this */
+
+                /* used by trees.c: */
+    /* Didn't use ct_data typedef below to suppress compiler warning */
+    struct ct_data_s dyn_ltree[HEAP_SIZE];   /* literal and length tree */
+    struct ct_data_s dyn_dtree[2*D_CODES+1]; /* distance tree */
```
Comment (jeongchanKim) [inline external/grpc/extlib/zlib/deflate.h]: add space between operator and operand

File: external/grpc/extlib/zlib/trees.h [hunk 9]
```diff
@@ -0,0 +1,128 @@
+/* header created automatically with -DGEN_TREES_H */
+
+local const ct_data static_ltree[L_CODES+2] = {
```
Comment (jeongchanKim) [inline external/grpc/extlib/zlib/trees.h]: add spaces

File: apps/examples/helloxx/Makefile [hunk 10]
```diff
@@ -1,6 +1,6 @@
 ###########################################################################
 #
-# Copyright 2016 Samsung Electronics All Rights Reserved.
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/helloxx/Makefile]: Revert this file changes

File: apps/examples/helloxx/helloxx_main.cxx [hunk 11]
```diff
@@ -1,6 +1,6 @@
 /****************************************************************************
  *
- * Copyright 2016 Samsung Electronics All Rights Reserved.
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/helloxx/helloxx_main.cxx]: revert

File: apps/examples/helloxx/helloxx_main.cxx [hunk 12]
```diff
@@ -18,7 +18,7 @@
 //***************************************************************************
 // examples/helloxx/helloxx_main.cxx
 //
-//   Copyright (C) 2009, 2011-2013 Gregory Nutt. All rights reserved.
+//   Copyright (C) 2009, 2011-2013, 2017 Gregory Nutt. All rights reserved.
```
Comment (sunghan-chang) [inline apps/examples/helloxx/helloxx_main.cxx]: revert

File: os/net/socket/sendmsg.c [hunk 13]
```diff
@@ -0,0 +1,104 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline os/net/socket/sendmsg.c]: 2018

File: os/net/socket/sendmsg.c [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+ssize_t sendmsg(int sockfd, struct msghdr *msg, int flags)
+{
+	uint8_t *buf = (uint8_t *)(msg->msg_iov->iov_base);
+	size_t len = msg->msg_iov->iov_len;
+	struct sockaddr *to = (struct sockaddr *)msg->msg_name;
+	int *addrlen = &(msg->msg_namelen);
+
+	//printf("\n[Received IOTIVITY Packet][%s:%d] \n", __FUNCTION__, __LINE__);
+
+	return sendto(sockfd, buf, len, flags, to, (socklen_t) *addrlen);
```
Comment (sunghan-chang) [inline os/net/socket/sendmsg.c]: violation of coding rule
remove a space after typecasting

File: apps/examples/helloxx/helloxx_main.cxx [hunk 15]
```diff
@@ -86,14 +87,14 @@
 #    define cxxvdbg           vdbg
 #    define cxxllvdbg         llvdbg
 #  else
-#    define cxxvdbg(x...)
-#    define cxxllvdbg(x...)
+#    define cxxvdbg(...)	printf(__VA_ARGS__)
```
Comment (sunghan-chang) [inline apps/examples/helloxx/helloxx_main.cxx]: Why do you set dbg as printf even though debug config is disabled?

File: apps/examples/protobuf/Make.defs [hunk 16]
```diff
@@ -0,0 +1,21 @@
+###########################################################################
+#
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/protobuf/Make.defs]: 2018. Please  check all of newly added files.

File: external/protobuf/.gitmodules [hunk 17]
```diff
@@ -0,0 +1,3 @@
+[submodule "third_party/benchmark"]
```
Comment (sunghan-chang) [inline external/protobuf/.gitmodules]: Is this submodule used?

**Comment by sunghan-chang** [comment]
The greeter and router client examples are running on gRPC only. But I don't know they are gRPC examples by name. How about grpc_greeter_client and grpc_router_client or put them under examples/grpc folder?

File: external/grpc/third_party/cares/ares_destroy.c [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+#include <assert.h>
+
+#include "ares.h"
+#include "ares_private.h"
+
+void ares_destroy_options(struct ares_options *options)
+{
+  int i;
+
+  if(options->servers)
```
Comment (jeongchanKim) [inline external/grpc/third_party/cares/ares_destroy.c]: add space between if and (

Comment (an4967) [reply external/grpc/third_party/cares/ares_destroy.c]: @jeongchanKim  We don't need to check the coding rule in external path

File: external/protobuf/.gitmodules (hunk 17 above)
Comment (an4967) [reply external/protobuf/.gitmodules]: @sunghan-chang This is not using now, but it will be used to check performance of protobuf later.

**Comment by an4967** [comment]
@sunghan-chang I changed the names of grpc examples as you commented

File: external/grpc/extlib/zlib/deflate.h (hunk 7 above)
Comment (pillip8282) [reply external/grpc/extlib/zlib/deflate.h]: it's external. so we don't need to follow the coding rule.

File: external/grpc/extlib/zlib/deflate.h (hunk 8 above)
Comment (pillip8282) [reply external/grpc/extlib/zlib/deflate.h]: it's external. so we don't need to follow the coding rule.

File: external/grpc/extlib/zlib/trees.h (hunk 9 above)
Comment (pillip8282) [reply external/grpc/extlib/zlib/trees.h]: it's external. so we don't need to follow the coding rule.

**Comment by sunghan-chang** [comment]
@lokeshbv Thank you for checking and notice. We are going to remove 143 line.

**Pull request merged**